[DON'T MERGE] Sync fixes

langerhans opened the pull request:
Not ready for merge yet, but PRing to have it here.  
You probably know better how you thought the block/header parsing will work, as my current solution feels kinda hacky.

This syncs a lot further than yesterday though, so it is progress. I won't be around until tomorrow afternoon/evening, so if you want to poke at it (or me), here it is.

367655d should be correct though, that's why it's an extra commit.

Currently it stops syncing at block `bd98a06391115285265c04984e8505229739f6ffa5d498929a91fbe7c281ea7b` with "Aux POW wrong index". I think then none may be wrong there.

rnicoll commented on src/main/java/org/bitcoinj/core/AuxPoW.java:
```diff
@@ -423,12 +423,13 @@ public Transaction getTransaction() {
      */
     static boolean arrayMatch(byte[] script, int offset, byte[] subArray) {
-        for (int matchIdx = 0; matchIdx + offset < script.length && matchIdx < subArray.length; matchIdx++) {
+        int matchIdx;
```
This is worth getting in quickly even while pondering the rest, could you make this and its corresponding test a separate PR I can merge ASAP please?

langerhans replied:
Opened #8

rnicoll commented on src/main/java/org/bitcoinj/core/AuxPowHeadersMessage.java:
```diff
@@ -0,0 +1,76 @@
+            length = 81 * (int)numHeaders;
+        }
+
+        long numHeaders = readVarInt();
```
Does this intentionally read the number of headers twice where an unknown length is provided?

rnicoll replied:
Yes, yes it does, ignore me :)

rnicoll commented:
https://github.com/rnicoll/bitcoinj/commit/04d04599fdbaf918cb02cffcfb7ed84ae3ea1b65 will form the basis of doing this less invasively, later, but doesn't actually work right now for reasons I can't see (checking the parsed headers by eye it looks fine)

rnicoll commented:
bitcoinj/bitcoinj#1102 should fix most of this

langerhans commented:
Updated this, still broken tho. I get to the same block, that fails to verify with the same Exception. I added the test, and under testing it parses and checks without issue. Something is fishy here.

The optimalEncodingSize change was needed cause parseTransactions would reset it to 80 since there are no transactions.  

Maybe you have an idea what's wrong :crystal_ball:

langerhans committed:
Advance the cursor the full header length

Otherweise parsing the nonexistant transactions would reset the size of the message to 80

langerhans commented:
Updated with the cursor length fix thing...
